src/widget/form/settings/avform.cpp

```diff
@@ -116,6 +117,12 @@ void AVForm::showEvent(QShowEvent* event)
     createVideoSurface();
     getVideoDevices();
 
+    // To don't block GUI thread.
+    QtConcurrent::run([this]()
+    {
+        camVideoSurface->setSource(&camera);
```

Comment:
@diadlo The actual problem here is: You're wrapping one problem in another one. However you can move the `QtConcurrent` call inside the setter and make that completely "self-managed", like this:

```c++
void VideoSurface::setSource(VideoSource *src)
{
    if (source == src)
        return;

    QtConcurrent::run([this]()
    {
        // note:
        // The actual problem is inside the (un-)subscribe methods,
        // which are long running operations.
        // It is a good idea to turn "source" into a `QExplicitlySharedDataPointer`, doing the
        // subscription in c'tor/d'tor of a shared object (inherit VideoSurface from `QSharedData`)
        unsubscribe();
        source = src;
        subscribe();
    }
}
``` 

In theory, calling a QWidget's method, that doesn't update the GUI, is safe. Creating a QWidget itself within thread context, other than the main thread (aka "GUI thread") leads to an assertion: `ASSERT failure in QWidget: "Widgets must be created in the GUI thread."` Why? Because rendering is not thread-safe and would cause a race, when updating the widget. Thus, it is not only bad practice, but it doesn't work at all.